File: core/src/net/brenig/pixelescape/game/data/GameDebugSettings.java
package net.brenig.pixelescape.game.data;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Preferences;

import net.brenig.pixelescape.PixelEscape;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Manages debug settings
 */
public class GameDebugSettings {


	private static final String PREF_DEBUG_SETTINGS = "PixelEscape_Debug_Preferences";

	private static final Map<String, Boolean> defaults = createDefaults();



	private final Preferences prefs;

	public GameDebugSettings() {
		prefs = Gdx.app.getPreferences(PREF_DEBUG_SETTINGS);
	}

	private static Map<String, Boolean> createDefaults() {
		Map<String, Boolean> result = new HashMap<String, Boolean>();
		result.put("SHOW_FPS", false);
		result.put("AUTO_PAUSE", true);
		result.put("PLAYER_EXPLOSION_RED", false);
		result.put("DEBUG_MODE_COORDS", false);
		result.put("DEBUG_WORLD_GEN_VALIDATE", false);
		result.put("DEBUG_SCREEN_BOUNDS", false);
		result.put("DEBUG_UI", false);
		result.put("DEBUG_LOGGING", false);
		result.put("DEBUG_GOD_MODE", false);
		result.put("DEBUG_MUSIC", false);
		result.put("SCREEN_SHAKE", true);
		return Collections.unmodifiableMap(result);
	}

	public void saveToDisk() {
		prefs.flush();
	}

	public boolean getBoolean(String s) {
		Boolean def = defaults.get(s);
		if(!PixelEscape.getPixelEscape().gameConfig.debugSettingsAvailable()) {
			//return default if debug settings are deactivated
			return def == null ? false : def;
		}
		if(def == null) {
			return prefs.getBoolean(s);
		}
		return prefs.getBoolean(s, def);
	}

	public void setBoolean(String s, boolean b) {
		prefs.putBoolean(s, b);
	}

	public static boolean get(String s) {
		return PixelEscape.getPixelEscape().gameDebugSettings.getBoolean(s);
	}

	public static void set(String s, boolean b) {
		PixelEscape.getPixelEscape().gameDebugSettings.setBoolean(s, b);
	}
}

File: core/src/net/brenig/pixelescape/game/data/GameDebugSettings.kt
package net.brenig.pixelescape.game.data

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Preferences
import net.brenig.pixelescape.PixelEscape
import java.util.*

/**
 * Manages debug settings
 */
class GameDebugSettings {
    private val prefs: Preferences

    init {
        prefs = Gdx.app.getPreferences(PREF_DEBUG_SETTINGS)
    }

    fun saveToDisk() {
        prefs.flush()
    }

    fun getBoolean(s: String): Boolean {
        val def = defaults[s]
        if (!PixelEscape.INSTANCE.gameConfig.debugSettingsAvailable) {
            //return default if debug settings are deactivated
            return def == true
        }
        return if (def == null) {
            prefs.getBoolean(s)
        } else prefs.getBoolean(s, def)
    }

    fun setBoolean(s: String, b: Boolean) {
        prefs.putBoolean(s, b)
    }

    companion object {
        private const val PREF_DEBUG_SETTINGS = "PixelEscape_Debug_Preferences"

        private val defaults = createDefaults()

        private fun createDefaults(): Map<String, Boolean> {
            val result = HashMap<String, Boolean>()
            result.put("SHOW_FPS", false)
            result.put("AUTO_PAUSE", true)
            result.put("DEBUG_MODE_COORDS", false)
            result.put("DEBUG_WORLD_GEN_VALIDATE", false)
            result.put("DEBUG_SCREEN_BOUNDS", false)
            result.put("DEBUG_UI", false)
            result.put("DEBUG_LOGGING", false)
            result.put("DEBUG_GOD_MODE", false)
            result.put("DEBUG_MUSIC", false)
            result.put("SCREEN_SHAKE", true)
            result.put("FORCE_TUTORIALS", false)
            return Collections.unmodifiableMap(result)
        }

        operator fun get(s: String): Boolean {
            return PixelEscape.INSTANCE.gameDebugSettings.getBoolean(s)
        }

        operator fun set(s: String, b: Boolean) {
            PixelEscape.INSTANCE.gameDebugSettings.setBoolean(s, b)
        }
    }
}
